src/chain/chain.rs

```diff
@@ -291,26 +291,33 @@ impl Chain {
         match event {
             AccumulatingEvent::SectionInfo(ref info)
             | AccumulatingEvent::NeighbourInfo(ref info) => {
-                let old_neighbours: BTreeSet<_> = self.neighbour_elders_p2p().cloned().collect();
+                let old_neighbours: BTreeMap<_, _> = self.neighbour_elder_map();
                 self.add_elders_info(info.clone(), proofs)?;
-                let new_neighbours: BTreeSet<_> = self.neighbour_elders_p2p().cloned().collect();
+                let new_neighbours: BTreeMap<_, _> = self.neighbour_elder_map();
 
                 if let Some((ref cached_info, _)) = self.state.split_cache {
                     if cached_info == info {
                         return Ok(None);
                     }
                 }
 
-                let neighbour_change = EldersChange {
-                    added: new_neighbours
-                        .difference(&old_neighbours)
-                        .cloned()
-                        .collect(),
-                    removed: old_neighbours
-                        .difference(&new_neighbours)
-                        .cloned()
-                        .collect(),
-                };
+                // Create BTreeSet for the purpose of taking the difference.
+                // WIP: can we do better?
+                let old_ids: BTreeSet<_> = old_neighbours.keys().cloned().collect();
+                let new_ids: BTreeSet<_> = new_neighbours.keys().cloned().collect();
+                let added_ids: BTreeSet<_> = new_ids.difference(&old_ids).cloned().collect();
+                let removed_ids: BTreeSet<_> = old_ids.difference(&new_ids).cloned().collect();
+
+                let added = new_neighbours
+                    .into_iter()
+                    .filter(|(pub_id, _)| added_ids.contains(pub_id))
+                    .collect();
```

Comment:
can remove new set and just do:
                 let added = new_neighbours
                    .into_iter()
                    .filter(|(pub_id, _)| !old_neighbours.contains_key(pub_id))
                    .collect();

Reply:
Ah good point, yes I'll try that! Thanks!